node/lib/util/merge_bare_util.js

```diff
@@ -0,0 +1,287 @@
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+"use strict";
+
+const assert       = require("chai").assert;
+const co           = require("co");
+const colors       = require("colors");
+const NodeGit      = require("nodegit");
+
+const CherryPickUtil      = require("./cherry_pick_util");
+const ConfigUtil          = require("./config_util");
+const DoWorkQueue         = require("./do_work_queue");
+const GitUtil             = require("./git_util");
+const Open                = require("./open");
+const UserError           = require("./user_error");
+
+
+/**
+ * Update meta repo index and point the submodule to a commit sha
+ * 
+ * @param {NodeGit.Index} index
+ * @param {String} subName
+ * @param {String} sha
+ */
+const addSubmoduleCommit = co.wrap(function *(index, subName, sha) {
+    assert.instanceOf(index, NodeGit.Index);
+    assert.isString(subName);
+    assert.isString(sha);
+
+    const entry = new NodeGit.IndexEntry();
+    entry.path = subName;
+    entry.mode = NodeGit.TreeEntry.FILEMODE.COMMIT;
+    entry.id = NodeGit.Oid.fromString(sha);
+    entry.flags = entry.flagsExtended = 0;
+    yield index.add(entry);
+});
+
+/**
+ * Merge in each submodule and update in memory index accordingly.
+ * 
+ * @param {NodeGit.Repository}  repo
+ * @param {Open.Opener}         opener
+ * @param {NodeGit.Index}       mergeIndex
+ * @param {Object}              subs  map from sub name to changes
+ * @param {String}              message commit message
+ * */
+exports.mergeSubmoduleBare = co.wrap(function *(repo, 
+                                                opener,
+                                                mergeIndex,
+                                                subs,
+                                                message) {
+    assert.instanceOf(repo, NodeGit.Repository);
+    assert.instanceOf(mergeIndex, NodeGit.Index);
+    assert.isObject(subs);
+    assert.isString(message);
+
+    const result = {
+        conflicts: {},
+        commits: {},
+    };
+    const sig = yield ConfigUtil.defaultSignature(repo);
+    const fetcher = yield opener.fetcher();
+
+    const mergeSubmodule = co.wrap(function *(name) {
+
+        const subRepo = yield opener.getSubrepo(name, true);
+        const change = subs[name];
+
+        const theirSha = change.newSha;
+        const ourSha = change.ourSha;
+        yield fetcher.fetchSha(subRepo, name, theirSha);
+        yield fetcher.fetchSha(subRepo, name, ourSha);
+        const theirCommit = yield subRepo.getCommit(theirSha);
+        const ourCommit = yield subRepo.getCommit(ourSha);
+
+        // No change if ours is up-to-date
```

Comment:
This check is from the submodule's perspective.  But the meta's perspective is what matters here.

Reply:
Just noticed that you hadn't gotten to this yet, so going to leave off reviewing for now -- let me know when you're ready for more.

Reply:
updated